odoo/models.py

```diff
@@ -6811,6 +6814,75 @@ def _invalidate_cache(self, fnames=None, ids=None):
                 spec.append((invf, None))
         self.env.cache.invalidate(spec)
 
+    @contextlib.contextmanager
+    def modifying_recordset(self, *, outputs: Iterable[str], inputs: Iterable[str] = (), discard_inputs: bool = False) -> Iterator[SimpleNamespace]:
+        """Helper for manipulating caches & computes around SQL queries.
+
+        Only affects the current recordset.
+
+        :param outputs: the list of fields modified by the SQL query, these
+            fields will be both :meth:`~.invalidate_recordset` and :meth:`~.modified`
+        :param inputs: the list of fields the SQL query needs / reads, these
+            fields will be :meth:`~.flush_recordset` unless `discard_inputs` is
+            set (in which case they are only invalidated)
+        :param discard_inputs: whether pending updates to the ``inputs`` should
+            be discarded / ignored, or applied before the query. Defaults to
+            ``False`` (any pending change to an input field will be flushed)
+        :return: a simple object whose `records` attribute can be updated to
+            control how many records are :meth:`~.modified`, useful if only a
+            subset of the recordset got updated
+        """
+        inputs = tuple(inputs)
+        outputs = tuple(outputs)
+        if discard_inputs:
+            self.invalidate_recordset(inputs + outputs, flush=False)
+        else:
+            if inputs:
+                self.invalidate_recordset(inputs, flush=True)
+            self.invalidate_recordset([f for f in outputs if f not in inputs], flush=False)
+
+        # alternative: raise a specific exception?
+        obj = SimpleNamespace(records=self)
+        yield obj
+        obj.records.modified(outputs)
```

Comment:
I can see quite a few issues here:
- What's the point of invalidating the inputs?  The inputs should be flushed, but unless they are also modified, dropping their value from memory doesn't solve any problem.  It actually makes things worse, since using those fields after the query will require to fetch them again.
- Invalidating the fields before yielding is correct _only if_ the body of the `with` statement doesn't access those fields before actually doing the query.  This requirement is _critical_ for the correctness of the abstraction, and it must be carefully documented.
- Calling `modified()` after the update is not enough.  When the modified fields actually modify relations (directly or indirectly by recomputing a relational field), then `modified()` should be called _before_ the update, too.  Changing a sales order line from one sales order to another one requires to recompute the total of _both_ the first sales order (that depends on the line before the update) and the second one (that depends on the line after the update).  See the implementation of method `write()`.
- In order to be fully consistent, one also needs to protect modified fields against recomputation.  If you modify both fields A and B, and B may be computed and depends on A, calling `modified()` without protecting that field will force the recomputation of B after its update.  Again, see method `write()`.
- Finally, you should also validate Python constraints with `_validate_fields()`.